force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,30 +246,53 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+         // Loop through each Contact
+        for (Contact con : contactsToUpsert) {
+            // Extract the Account Name from Contact's LastName
+            String accountName = con.LastName;
+        
+            // Get or create an Account with the extracted name
+            Account acc = upsertAccount(accountName);
+        
+            // Link the Contact to its Account using the Account Id
+            con.AccountId = acc.Id;
+        }
```

Comment:
`upsertAccount` does SOQL + DML within a for loop. What happens if we have 200 contactsToUpsert? 

You'll need to find a way to use a map and a single query + single DML statements outside of loops to solve this problem.